Experimental support of Airflow as a runtime

akchinSTC opened the pull request:
Implements experimental support for Airflow runtime

Fixes #185 
Fixes #1183
Fixes #1278 


Developer's Certificate of Origin 1.1

       By making a contribution to this project, I certify that:

       (a) The contribution was created in whole or in part by me and I
           have the right to submit it under the Apache License 2.0; or

       (b) The contribution is based upon previous work that, to the best
           of my knowledge, is covered under an appropriate open source
           license and I have the right under that license to submit that
           work with modifications, whether created in whole or in part
           by me, under the same open source license (unless I am
           permitted to submit under a different license), as indicated
           in the file; or

       (c) The contribution was provided directly to me by some other
           person who certified (a), (b) or (c) and I have not modified
           it.

       (d) I understand and agree that this project and the contribution
           are public and that a record of the contribution (including all
           personal information I submit with it, including my sign-off) is
           maintained indefinitely and may be redistributed consistent with
           this project or the open source license(s) involved.

kevin-bates commented on elyra/pipeline/processor_airflow.py:
```diff
@@ -0,0 +1,243 @@
+        :return: metadata in json format
+        """
+        try:
+            runtime_configuration = MetadataManager(namespace="runtime").get(name)
```
Namespace is now plural: `runtimes`.

kevin-bates commented on elyra/pipeline/processor_airflow.py:
```diff
@@ -0,0 +1,243 @@
+        """
+        try:
+            runtime_configuration = MetadataManager(namespace="runtime").get(name)
+            return runtime_configuration
```
I recommend moving the return statement to end of the method.

kevin-bates commented on elyra/pipeline/airflow_template.py:
```diff
@@ -0,0 +1,101 @@
+    # Default arguments to use with Operator
+    default_args = {'owner': 'elyra',
+                    'depends_on_past': False,
+                    'start_date': dates.days_ago(1),
```
I find this a little confusing, even after googling it (and, to my relief, others do as well).  Would it be possible to either add a comment to this line, or extend the comment on L34 to perhaps indicate what "1 day ago" means?  I suspect it's just how you trigger an immediate schedule dynamically.

kevin-bates commented on elyra/templates/airflow/notebook_op.py:
```diff
@@ -54,18 +51,19 @@ def __init__(self,
 
         if 'image' not in kwargs:  # default image used if none specified
-            kwargs['image'] = 'lresende/notebook-kubeflow-pipeline:dev'
+            kwargs['image'] = 'elyra/tensorflow:1.15.2-py3'
```
Once Luciano's BYOImage PR lands, we'll probably want to way to somehow convey that a given image is "the default".  If they're loaded in a well-defined order, this could be as simple as taking the first in the list, or we could add a property that indicates the instance can be used as a default (but we'd likely need to enforce uniqueness across the instances (so only one is the default)). At any rate, nothing to do now, but we'll eventually need to pull this value from somewhere.

kevin-bates commented on elyra/pipeline/processor_airflow.py:
```diff
@@ -39,30 +38,25 @@ def type(self):
-        airflow_client.trigger_dag(dag_id='elyra', run_id=pipeline_name, conf=payload)
-
-        self.log.info("Starting AirFlow Pipeline Run...")
+        self.log.info("Uploaded AirFlow Pipeline...please wait for Airflow Scheduler to start a run")
```
Since this is going to a log file, let's change "please wait" to "waiting".

akchinSTC commented:
Updated with suggestions from @kevin-bates
 
Working on dropping the DAGs into an S3 bucket and syncing that up with the dag bag.

kevin-bates commented on elyra/pipeline/processor_airflow.py:
```diff
@@ -39,30 +38,25 @@ def type(self):
-        timestamp = datetime.now().strftime("%m%d%H%M%S")
-        pipeline_name = (pipeline.title if pipeline.title else 'pipeline') + '-' + timestamp
+        unique_id = uuid.uuid4().hex[4:15]
+        pipeline.title = pipeline.title + '-' + unique_id
```
The previous code only applied the uniqueness if title wasn't set.  This makes it unconditional and, if not set, will result in a title like '-9347a3c323f'.  Does pipeline.title always have a value prior to this code?

I like the idea of applying uniqueness unconditionally, just want to be sure if we still need the check for title not having a value or not.

kevin-bates commented on elyra/pipeline/processor_airflow.py:
```diff
@@ -15,19 +15,18 @@
 from elyra.util.cos import CosClient
-from airflow.api.client.json_client import Client as AirflowClient
-from urllib3.exceptions import MaxRetryError
+# from airflow.api.client.json_client import Client as AirflowClient
```
If never to be used, we should remove.

kevin-bates commented on elyra/pipeline/processor_airflow.py:
```diff
@@ -39,30 +38,25 @@ def type(self):
 
         # TODO update pipeline json to flow runtime type
-        runtime_configuration = self._get_runtime_configuration(pipeline.runtime_config)
+#        runtime_configuration = self._get_runtime_configuration(pipeline.runtime_config)
```
ditto if never to be used (and next statement as well).

kevin-bates commented on elyra/pipeline/processor_airflow.py:
```diff
@@ -0,0 +1,254 @@
+        if not pipeline.title:
+            pipeline_name = "pipeline" + '-' + unique_id
+        else:
+            pipeline_name = pipeline.title + '-' + unique_id
```
For another time, but this logic should be baked into the Pipeline class via something like a `name` or 'run_name` property.

kevin-bates commented on elyra/pipeline/processor_airflow.py:
```diff
@@ -0,0 +1,254 @@
+
+        return pipeline_export_path
+
+    def _artifact_list_to_str(self, pipeline_array):
```
All of the following methods appear to be identical to those in [`KfpPipelineProcessor`](https://github.com/elyra-ai/elyra/blob/master/elyra/pipeline/processor_kfp.py#L220-L259) and should be moved to `PipelineProcessor`.

kevin-bates commented:
For the future, I think the code that's in the processors should move into the respective Pipeline classes and only have a single PipelineProcessor class that acts on a well-defined interface implemented by the Pipeline classes.  This would enable a single set of pipeline-specific class hierarchies so folks that want to "bring their own pipeline" would simply implement one class.  (At least that's my initial observation from seeing the kinds of changes that adding a second pipeline introduces.). I would be happy to help with designing that object model.

akchinSTC commented:
Rebased, updated with some recommended changes, updated jinja template to include start on upload option so we don't have to mess with airflows global config and added some documentation to start. Will need to decide on how to install the Notebook Operator

toby-coleman commented:
Hi - I'm new to Elyra, and really like what I've seen so far. Wondered when this PR might get merged, as I'd love to try out the Airflow runtime.

akchinSTC commented:
@toby-coleman, Thanks for the interest!
This PR is a bit stale atm and needs to be rebased since a lot of changes in the pipeline processors have been made. The underlying principle is the same though of submitting the generated DAG to a shared dag bag either in a FS or object storage. Unfortunately we are bit busy right now with other tasks, but if you want to take a stab at this I would be more than happy to answer any questions you might have.

toby-coleman commented:
@akchinSTC thanks for getting back to me.  I've had a look through the code in the original PR to see what I'd have to do to bring it up-to-date.  One question: in the [airflow template](https://github.com/akchinSTC/elyra-1/blob/akchin-airflow-05041600/elyra/templates/airflow/airflow_template.jinja2) there's an import for the Airflow operator that executes each pipeline step: `from notebook.pipeline import NotebookOp`

I couldn't find the code for this operator - is it in a separate repo somewhere?

akchinSTC commented:
@bourdakos1 - Hi Nick, with this PR when I export an Airflow pipeline I only get the following two options. What would be the best way for me to go about adding a third option, or perhaps a separate list options depending on the type of runtime selected. This will probably lean into the work you are doing for pipelines 2.0 I think. Thanks in advance for helping me take a look. 
![image](https://user-images.githubusercontent.com/13340013/98052317-bae42a00-1dea-11eb-9670-5626f15c6fb0.png)

bourdakos1 commented:
@akchinSTC I think getting a list of export options will be part of the API, but right now they are hard coded. 

For now you should be able to just add an option to this list:
https://github.com/elyra-ai/elyra/blob/60618c00402a7908a08fb2067ab37931040d2ca7/packages/pipeline-editor/src/PipelineExportDialog.tsx#L21

Not sure what you would need to do on the backend though

kevin-bates reviewed (commented):
Aside from the explicit comments, it seems like all of the updates to the front-end packages "package" are _replacing_ 'KFP' with 'AIRFLOW' and I don't understand what that is about.

akchinSTC commented on elyra/util/cos.py:
```diff
@@ -72,8 +72,10 @@ def upload_file(self, file_name, file_path):
         :return:
         """
+        bucket = self.bucket
+
```
corrected

akchinSTC commented on elyra/pipeline/processor_airflow.py:
```diff
@@ -0,0 +1,314 @@
+        return name + '-' + operation.id + ".tar.gz"
+
+    def _get_dependency_source_dir(self, operation):
+        return os.path.join(os.getcwd(), os.path.dirname(operation.filename))
```
removed with self.root_dir being used in new subclass

akchinSTC commented on packages/application/src/test/application.spec.ts:
```diff
@@ -99,7 +99,7 @@ describe('@elyra/application', () => {
           return schema.name;
         });
-        const knownSchemaNames = ['code-snippet', 'runtime-image', 'kfp'];
+        const knownSchemaNames = ['code-snippet', 'runtime-image', 'airflow'];
```
added instead of replaced

akchinSTC commented on elyra/pipeline/processor_airflow.py:
```diff
@@ -0,0 +1,314 @@
+
+        return archive_artifact
+
+    def _get_runtime_configuration(self, name):
```
new subclass created

karlaspuldaro commented:
Working on each dialog at a time. This is how the export one currently looks like:
![export-dialog](https://user-images.githubusercontent.com/25207344/106818624-8b8dc500-6646-11eb-8c28-0ad83a1a46bb.gif)

ptitzler commented:
@karlaspuldaro Could you please change the `Runtime Config` label text to `Runtime Configuration`?

kevin-bates commented on docs/source/user_guide/runtime-conf.md:
```diff
@@ -166,9 +166,10 @@ Example: `['test','unsecured']`
 The KubeFlow Pipelines API endpoint you want to utilize. This setting is required.
 
-Example: `https://kubernetes-service.ibm.com/pipeline`
+KubeFlow Example: `https://kubernetes-service.ibm.com/pipeline`
```
I don't think we need to call out that this particular example is Kubeflow given we're clearly in the Kubeflow Settings section.  We also don't prefix `Example` in other examples.
```suggestion
Example: `https://kubernetes-service.ibm.com/pipeline`
```

akchinSTC replied:
done

kevin-bates commented on docs/source/user_guide/runtime-conf.md:
```diff
@@ -198,9 +199,39 @@ kubectl describe configmap -n kubeflow kfp-tekton-config
+
+The Apache Airflow API endpoint you want to utilize. This setting is required.
+
+Airflow Example: `http://your.apache.webserver:port`
```
Same here - no need to call out this as special to Airflow.

akchinSTC replied:
done

karlaspuldaro commented:
Notebook submission dialog
Before:
![image](https://user-images.githubusercontent.com/25207344/107429115-925a8300-6af1-11eb-9615-71aecf4dead0.png)

After:
![image](https://user-images.githubusercontent.com/25207344/107429356-ccc42000-6af1-11eb-8753-c467d590066f.png)

lresende reviewed (approved):
Overall it LGTM for merging into master, and then I would suggest we take a week or two for folks to validate, test, stabilize and review the docs. Thoughts?

akchinSTC commented:
@lresende - I am good with merging into master. We can shake out the issues as more people use it.

ajbozarth commented:
Work on https://github.com/elyra-ai/elyra/issues/1278 will continue in a follow up PR so this can be merged.

Once this is merged, until that Issue is addressed, it will be impossible to create a new KFP runtime via the UI. Creating kip runtimes will still; be possible via other methods and editing existing kip runtimes still works.

lresende commented:
All CI tests are green, merging. 

Again, Thanks @akchinSTC for all the work here...
